File: test_drive/lib/widgets/hybrid_editor/core/transactions/update_list_level_operation.dart
import 'package:test_drive/widgets/hybrid_editor/core/document.dart';
import 'package:test_drive/widgets/hybrid_editor/core/transactions/operation.dart';

class UpdateListLevelOperation implements Operation {
  final int blockIndex;
  final int oldLevel;
  final int newLevel;

  const UpdateListLevelOperation({
    required this.blockIndex,
    required this.oldLevel,
    required this.newLevel,
  });

  @override
  OperationType get type => OperationType.updateListLevel;

  @override
  Document apply(Document document) {
    final block = document[blockIndex];
    return document.updateBlock(
      blockIndex,
      block.copyWith(listLevel: newLevel),
    );
  }

  @override
  Operation inverse(Document document) {
    return UpdateListLevelOperation(
      blockIndex: blockIndex,
      oldLevel: newLevel,
      newLevel: oldLevel,
    );
  }
}

File: components/editor/core/operations/UpdateListLevelOperation.ts
import { copyBlock } from "../BlockNode";
import { type Document, updateBlock } from "../Document";
import { type Operation, OperationType } from "./Operation";

/// Operation to update the list level of a block
export class UpdateListLevelOperation implements Operation {
	readonly type = OperationType.updateListLevel;

	constructor(
		readonly blockIndex: number,
		readonly oldLevel: number,
		readonly newLevel: number,
	) {}

	apply(document: Document): Document {
		const block = document.blocks[this.blockIndex];
		return updateBlock(
			document,
			this.blockIndex,
			copyBlock(block, { listLevel: this.newLevel }),
		);
	}

	inverse(document: Document): Operation {
		return new UpdateListLevelOperation(
			this.blockIndex,
			this.newLevel,
			this.oldLevel,
		);
	}
}
